prow/spyglass/viewers/viewers.go

```diff
@@ -0,0 +1,201 @@
+	UseContext(ctx context.Context) error
+}
+
+// ViewHandler consumes artifacts and some possible callback json data and returns an html view.
+// Use the javascript function refreshView(viewName, viewData) to allow your viewer to call back to itself
+// (request more data, update the view, etc.). ViewData is a json blob that will be passed back to the
+// handler function for your view as the string
+type ViewHandler func([]Artifact, string) string
+
+// View gets the updated view from an artifact viewer with the provided name
+func View(name string, artifacts []Artifact, raw string) (string, error) {
+	handler, ok := viewHandlerRegistry[name]
+	if !ok {
+		return "", ErrInvalidViewName
+	}
+	return handler(artifacts, raw), nil
+
+}
+
+// Title gets the title of the view with the given name
+func Title(name string) (string, error) {
+	m, ok := viewMetadataRegistry[name]
+	if !ok {
+		return "", ErrInvalidViewName
+	}
+	return m.Title, nil
+
+}
+
+// Priority gets the priority of the view with the given name
+func Priority(name string) (int, error) {
+	m, ok := viewMetadataRegistry[name]
+	if !ok {
+		return -1, ErrInvalidViewName
+	}
+	return m.Priority, nil
+
+}
+
+// RegisterViewer registers new viewers
+func RegisterViewer(viewerName string, metadata ViewMetadata, handler ViewHandler) error {
+	_, ok := viewHandlerRegistry[viewerName]
+	if ok {
+		return fmt.Errorf("viewer already registered with name %s", viewerName)
+	}
+
+	if metadata.Title == "" {
+		return errors.New("empty title field in view metadata")
+	}
+	if metadata.Priority < 0 {
+		return errors.New("priority must be greater than 0")
+	}
+	viewHandlerRegistry[viewerName] = handler
+	viewMetadataRegistry[viewerName] = metadata
+	logrus.Infof("Spyglass registered viewer %s with title %s.", viewerName, metadata.Title)
+	return nil
+}
+
+// UnregisterViewer unregisters viewers
+func UnregisterViewer(viewerName string) error {
+	_, ok := viewHandlerRegistry[viewerName]
+	if !ok {
+		return ErrInvalidViewName
+	}
+	delete(viewHandlerRegistry, viewerName)
+	delete(viewMetadataRegistry, viewerName)
+	logrus.Infof("Spyglass unregistered viewer %s.", viewerName)
+	return nil
+}
+
+// LastNLines reads the last n lines from an artifact.
+func LastNLines(a Artifact, n int64) ([]string, error) {
+	// 300B, a reasonable log line length, probably a bit more scalable than a hard-coded value
+	return LastNLinesChunked(a, n, 300*n+1)
+}
+
+// LastNLinesChunked reads the last n lines from an artifact by reading chunks of size chunkSize
+// from the end of the artifact. Best performance is achieved by:
+// argmin 0<chunkSize<INTMAX, f(chunkSize) = chunkSize - n * avgLineLength
+func LastNLinesChunked(a Artifact, n, chunkSize int64) ([]string, error) {
+	toRead := chunkSize + 1 // Add 1 for exclusive upper bound read range
+	chunks := int64(1)
+	var contents []byte
+	var linesInContents int64
+	artifactSize, err := a.Size()
+	if err != nil {
+		return []string{}, err
```

Comment:
I would expect `return nil, err` in these kinds of situations (and please own your own errors)